### ISSUE TITLE
Implement something similar to action class in selenium for easier use of perform action

### ISSUE THREAD

**Issue by praveendvd** [webdriverio/webdriverio#7213]
**Is your feature request related to a problem? Please describe.**

Currently perform action requires knowledge on w3c action pragmatic implementation

**Describe the solution you'd like**

It would be great to wrap these actions in something like builder pattern like in selenium

`await browser.chainActions().moveTo(2,3).pause(1).moveTo(2,3).doubleClick().build()`

**Describe alternatives you've considered**

creating my own wrap like

```
moveTo: ()=> { 
actionArry.push(the action object
return this
)

build: async () => await browser.performActions(actionArr)
```

**Comment by christian-bromann** [comment]
What API would you suggest?

**Comment by praveendvd** [comment]
> What API would you suggest?

Like a chain action class that exposes all available actions as method , and finally a build step that triggers perform actions as shown in my example .

**Comment by praveendvd** [comment]
> What API would you suggest?

In gitter we have lots questions on this , thought having something like this can make it easier

**Comment by erwinheitzman** [comment]
I have taken a small example and modified it a little to match what it could look like for us.
This reminds me of how Protractor implemented the Actions as well and it felt very easy to use. It would also allow us to assign proper typings so that people don't make mistakes which makes me excited 😃 
```js
const actions = new Actions();
actions.moveToElement(element).click().perform();
```

**Comment by erwinheitzman** [comment]
This would also mean that our `dragAndDrop` and `keys` command and probably more would become part of this instead as you would have a single entity for performing actions like `clickAndHold` `contextClick` `doubleClick` `dragAndDrop` `keyDown` `keyUp` etc

**Comment by praveendvd** [comment]
> This would also mean that our `dragAndDrop` and `keys` command and probably more would become part of this instead as you would have a single entity for performing actions like `clickAndHold` `contextClick` `doubleClick` `dragAndDrop` `keyDown` `keyUp` etc

do we need to create action object? instead we can do like broser method , else it will be an entirely different implementation than other things . i think its better to extend the browser to have chainaction method

**Comment by erwinheitzman** [comment]
I'm currently trying a POC as we speak and even though I felt this would be handy I now feel that we could have the following:
```js
browser
  .click() // would perform click at cursor location
  .contextClick() // would perform context-click at cursor location
```
with
```js
$('some selector')
  .click() // would perform click at the center of the element
  .contextClick() // would perform context-click at the center of the element
```

What do you think about that @praveendvd?

**Comment by praveendvd** [comment]
> I'm currently trying a POC as we speak and even though I felt this would be handy I now feel that we could have the following:
> ```js
> browser.click() // would perform click at cursor location
> browser.contextClick() // would perform context-click at cursor location
> ```
> with
> ```js
> $('some selector').click() // would perform click at the center of the element
> $('some selector').contextClick() // would perform context-click at the center of the element
> ```
> 
> What do you think about that @praveendvd?

That sounds great , but I am not sure if there is a webdriver api to retrieve current cursor location 

Also having a chainable method helps in more complicated actions 

So adding browser.click() or DoubleClick etc to click at current location will be really useful , but that won't be a replacement for chainableaction I believe

**Comment by erwinheitzman** [comment]
Created a working POC as an example. Let me know what you think.

Steps:
- `npx wdio /path/to/poc`
- select typescript and accept defaults for everything else
- paste the below code in the test example (replace everything in the `it`)
- run `npx wdio ./wdio.conf.ts`

```js
        class Actions {
            actions: any[] = []
        
            click() {
                this.actions.push([{
                    type: 'pointer',
                    id: 'click',
                    parameters: { pointerType: 'mouse' },
                    actions: [
                        { type: 'pointerDown', button: 0 },
                        { type: 'pointerUp', button: 0 }
                    ]
                }])
                return this
            }
        
            clickAndHold() {
                this.actions.push([{
                    type: 'pointer',
                    id: 'clickAndHold',
                    parameters: { pointerType: 'mouse' },
                    actions: [
                        { type: 'pointerDown', button: 0 },
                    ]
                }])
                return this
            }
        
            contentClick() {
                this.actions.push([{
                    type: 'pointer',
                    id: 'contentClick',
                    parameters: { pointerType: 'mouse' },
                    actions: [
                        { type: 'pointerDown', button: 2 },
                        { type: 'pointerUp', button: 2 }
                    ]
                }])
                return this
            }
        
            moveToElement(element: any) {
                this.actions.push([{
                    type: 'pointer',
                    id: 'moveToElement',
                    parameters: { pointerType: 'mouse' },
                    actions: [{ type: 'pointerMove', origin: element, x: 0, y: 0 }]
                }])
                return this;
            }
        
            async perform() {
                for (const actionSet of this.actions) {
                    await browser.performActions(actionSet)
                }

                await browser.releaseActions()
            }
        }

        loginPage.open();

        const actions = new Actions();

        await actions
            .moveToElement(await $('#password'))
            .contentClick()
            .perform();

        await browser.pause(5000);
```

**Comment by erwinheitzman** [comment]
See #7214 for the POC as a Draft PR

**Comment by christian-bromann** [comment]
@erwinheitzman I find it difficult to use a class to initiate an action chain. We don't do this anywhere else. I would rather prefer something like:

```js
        const actions = browser.actions();

        await actions
            .moveToElement(await $('#password'))
            .contentClick()
            .perform();

        await browser.pause(5000);
```

My next question is, can we define which methods such an action class should expose?

**Comment by praveendvd** [comment]
> @erwinheitzman I find it difficult to use a class to initiate an action chain. We don't do this anywhere else. I would rather prefer something like:
> 
> ```js
>         const actions = browser.actions();
> 
>         await actions
>             .moveToElement(await $('#password'))
>             .contentClick()
>             .perform();
> 
>         await browser.pause(5000);
> ```
> 
> My next question is, can we define which methods such an action class should expose? Also I would like to suggest that this command and all actions we define need to work for web as well as for mobile.
@christian-bromann   I agree , we should not have a need of initializing a class .
i think we should be able to pass the pointer type to actions() like browser.actions('touch').moveTo etc to make it work with different devices?

**Comment by christian-bromann** [comment]
We can detect mobile sessions using `browser.isMobile`. I would suggest to make it so that you can use the same code for web actions than for mobile native actions.

**Comment by erwinheitzman** [comment]
I was thinking about implementing it as `browser.actions` later on but for the sake of local testing I have put it in a class for now. I'll work some on this to see if I can get something nice for both web and mobile.

Can we use the `performActions` for mobile or is that not possible? I'm honestly not sure what requirements mobile has so if anyone could help me with that that would be great :)

**Comment by christian-bromann** [comment]
> Can we use the `performActions` for mobile or is that not possible? I'm honestly not sure what requirements mobile has so if anyone could help me with that that would be great :)

For mobile we can use `touchAction` which is similar to the `performAction` command.

**Comment by erwinheitzman** [comment]
I found out that the `touch` is actually supported by now so we can remove the check (forgot where it is but I will fix that) and use this safely it seems.

Currently it works like this:
```js
await (await browser.actions())
  .moveToElement()
  .click()
  .perform();
```

How would I be about to get `actions` as a property of the Browser object that I would not have to await?

**Comment by christian-bromann** [comment]
> How can I implement `actions` as a `getter` method / property of the Browser object that I would not have to await?

You will need to make use of [Proxies](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy).

@erwinheitzman before you move on, can we quickly spec out the whole interface and make sure to agree on it?

**Comment by erwinheitzman** [comment]
@christian-bromann at the moment the code is still used like so:

```js
    await (await browser.actions())
      .moveToElement(await $("#password"))
      .perform()
```

Could you help out in some way to change this as I have no idea where and how to apply the Proxy?

**Comment by praveendvd** [comment]
> Would it be also possible (preferable before we continue with any implementations) to define which commands should exist and what their parameters should be? It would help us make sure we find the optimal interface. For example I would argue that something like `contextClick` is not necessary and would prefer something like `.click({ type: 'left' | 'middle' | 'right' })`. Also instead of `pressAndHold` I would prefer `.mousedown().pause(5000)`, seems for me more intuitive otherwise we will end up having more commands like e.g. pressAndHoldAndClickAndFoo.

Yes I agree some methods could be replaced with arguments.

**Comment by erwinheitzman** [comment]
I personally prefer a `click`, `middleClick` and `rightClick` in that case for convenience without having to pass in any options.

Regarding the `pressAndHold` action (apologies here that should have been `clickAndHold`), does not have a timeout. It presses down something until you tell it to release, so I disagree here.
Also tho whole idea of this Actions class is to chain commands, so I don't see something like `pressAndHoldAndClickAndFoo` happening but rather the following (which is exactly how the `dragAndDrop` works):
```js
browser.actions()
  .clickAndHold(await $('some-selector'))
  .pause(10)
  .moveToElement(await $('a-different-selector'))
  .release()
  .perform()
```

[Reference](https://github.com/webdriverio/webdriverio/pull/7214/files#diff-0ed18c34904e09e0002fcd7ec017cd2d740a1d78a3d7d6de34be073a44e9f452R380)

**Comment by christian-bromann** [comment]
Implemented by #8595
